File: nodejs/recognize.js
const fs = require('fs');
const {createSttClient, forEachRecognitionResult, checkWavFormat} = require('./common.js');
const {buildRecognizeCommand} = require('./args.js');
const stream = require('stream');
const wav = require('wav');
const concat = require('concat-stream');
const util = require('util');


const argv = buildRecognizeCommand().parse();

function main() {
    const sttClient = createSttClient();
    const reader = fs.createReadStream(argv.file);
    reader.on('error', (err) => {
        console.error(err);
        process.exit(1);
    });
    const concatStream = concat((buf) => {
        const unaryCall = sttClient.Recognize({
            config: {
                encoding: argv.encoding,
                sampleRateHertz: argv.rate,
                languageCode: argv.languageCode,
                maxAlternatives: argv.maxAlternatives,
                numChannels: argv.numChannels,
                enableAutomaticPunctuation: argv.automaticPunctuation,
                doNotPerformVad: (argv.performVad ? undefined : true),
                vadConfig: (!argv.performVad ? undefined : {
                    silenceDurationThreshold: argv.silenceDurationThreshold
                }),
            },
            audio: {
                content: buf,
            }
        }, (error, response) => {
            if (error != null) {
                console.error(error);
            } else {
                console.log(util.inspect(response, false, null, true));
            }
        });
        unaryCall.on('metadata', (metadata) => console.log("Initial response metadata", metadata));
        unaryCall.on('status', (status) => {
            console.log("Call ended, status", status);
            sttClient.close();
        });
    });

    if (argv.file.endsWith(".wav")) {
        const wavReader = wav.Reader();
        wavReader.on('format', (format) => {
            if (!checkWavFormat(format, argv)) {
                process.exit(1);
            } else {
                wavReader.pipe(concatStream);
            }
        });
        reader.pipe(wavReader);
    } else {
        reader.pipe(concatStream);
    }
}

main();

File: golang/cmd/recognize/recognize.go
package main

import (
	"context"
	"github.com/TinkoffCreditSystems/voicekit-examples/golang/pkg/args"
	"github.com/TinkoffCreditSystems/voicekit-examples/golang/pkg/common"
	sttPb "github.com/TinkoffCreditSystems/voicekit-examples/golang/pkg/tinkoff/cloud/stt/v1"
	"io"
	"io/ioutil"
	"os"
	"strings"
)

func main() {
	opts := args.ParseRecognizeOptions()
	if opts == nil {
		os.Exit(1)
	}
	defer opts.InputFile.Close()

	var dataReader io.Reader
	if strings.HasSuffix(opts.InputFile.Name(), ".wav") {
		reader, err := common.OpenWavFormat(opts.InputFile, *opts.Encoding, *opts.NumChannels, *opts.Rate)
		if err != nil {
			panic(err)
		}
		dataReader = reader
	} else {
		dataReader = opts.InputFile
	}

	client, err := common.NewSttClient()
	if err != nil {
		panic(err)
	}
	defer client.Close()

	contents, err := ioutil.ReadAll(dataReader)
	if err != nil {
		panic(err)
	}

	request := &sttPb.RecognizeRequest{
		Config: &sttPb.RecognitionConfig{
			Encoding:                   sttPb.AudioEncoding(sttPb.AudioEncoding_value[*opts.Encoding]),
			SampleRateHertz:            uint32(*opts.Rate),
			LanguageCode:               *opts.LanguageCode,
			MaxAlternatives:            uint32(*opts.MaxAlternatives),
			ProfanityFilter:            false,
			EnableAutomaticPunctuation: !(*opts.DisableAutomaticPunctuation),
			NumChannels:                uint32(*opts.NumChannels),
		},
		Audio: &sttPb.RecognitionAudio{
			AudioSource: &sttPb.RecognitionAudio_Content{Content: contents},
		},
	}
	if *opts.DoNotPerformVad {
		request.Config.Vad = &sttPb.RecognitionConfig_DoNotPerformVad{DoNotPerformVad: true}
	} else {
		request.Config.Vad = &sttPb.RecognitionConfig_VadConfig{
			VadConfig: &sttPb.VoiceActivityDetectionConfig{
				SilenceDurationThreshold: float32(*opts.SilenceDurationThreshold),
			},
		}
	}

	// NOTE: in production code you should probably use context.WithCancel, context.WithDeadline or context.WithTimeout
	// instead of context.Background()
	result, err := client.Recognize(context.Background(), request)
	if err != nil {
		panic(err)
	}
	if common.PrettyPrintProtobuf(result) != nil {
		panic(err)
	}
}
